### ISSUE TITLE
Counting of non-countable objects

### ISSUE THREAD

**Issue by rxu** [twigphp/Twig#2382]
An early look: in upcoming PHP 7.2, the RFC has been implemented https://wiki.php.net/rfc/counting_non_countables
Since then, calling ```count()``` with a parameter that is a scalar, null, or an object that doesn't implement Countable raises PHP warning like ```Warning: count(): Parameter must be an array or an object that implements Countable in %s on line %d```.

Thus, since PHP 7.2 this will be considered as a bug in the code.

Examples: 
https://github.com/twigphp/Twig/blob/2.x/lib/Twig/Node/Module.php#L370
https://github.com/twigphp/Twig/blob/2.x/lib/Twig/Extension/Core.php#L1117

and other places.

**Comment by stof** [comment]
Well, the length filter of Twig is meant to be used on strings and on countable values (array of Countable objects). So this is not a bug in Twig (it may be a bug in your own PHP code if you try to count something uncountable, but then the PHP warning is actually helpful to detect that you have a bug).

And Twig_Node_Module does not have a bug. ``$node`` is a Twig_Node instance, which implements Countable.

**Comment by rxu** [comment]
> the length filter of Twig is meant to be used on strings and on countable values

Actually, using ```count()``` on strings is not allowed in PHP 7.2 and raising warning as well.

**Comment by xabbuh** [comment]
@rxu The case of strings is covered by the `is_scalar()` check inside the implementation of the `length` filter which then does not execute `count()`, but determines the length of the scalar's string representation.

**Comment by rxu** [comment]
Good, but what about null values?

**Comment by stof** [comment]
why would you count null values in Twig ? This is a case where the PHP warning would actually help you detect a bug in your own code

**Comment by rxu** [comment]
F.e. ```{% if var|length >3 %}``` while var is dynamic and not set yet.

**Comment by xabbuh** [comment]
I would then use the `default` filter like this:

```twig
{% if var|default([])|length > 3 %}
    {# ... #}
{% endif %}
```

**Comment by fabpot** [comment]
Closing as @xabbuh explained how to do that.

**Comment by PatricNox** [comment]
_Necro_

if anyone gets here through google, like me, and thought there was no fix for this: upgrade to 3.0.1 or higher.

https://github.com/twigphp/Twig/commit/d2b7a01a181efa9ff081df1e43226605e33bed8d#diff-d7378002f67a61c458c1de0468eef74f
